pkg/sentry/fsimpl/fuse/init.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+)
+
+// consts used by FUSE_INIT negotiation.
+const (
+	// fuseMaxMaxPages is the maximum value for MaxPages received in InitOut.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMaxMaxPages = 256
+
+	// Maximum value for the time granularity for file time stamps, 1s.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMaxTimeGranNs = 1000000000
+
+	// Minimum value for MaxWrite.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMinMaxWrite = 4096
+
+	// Temporary default value for max readahead, 128kb.
+	fuseDefaultMaxReadahead = 131072
+
+	// The FUSE_INIT_IN flags sent to the daemon.
+	// TODO(gvisor.dev/issue/3199): complete the flags.
+	fuseDefaultInitFlags = linux.FUSE_MAX_PAGES
+)
+
+// Adjustable maximums for Connection's cogestion control parameters.
+// Used as the upperbound of the config values.
+// Currently we do not support adjustment to them.
+var (
+	MaxUserBackgroundRequest   uint16 = fuseDefaultMaxBackground
+	MaxUserCongestionThreshold uint16 = fuseDefaultCongestionThreshold
+)
+
+// InitSend sends a FUSE_INIT request.
+func (conn *connection) InitSend(creds *auth.Credentials, pid uint32) error {
```

Comment:
I don't love that these are methods on the connection. The Connection is used by all operators to talk to the server. Namely the methods it should export are the Call methods, the request methods, and the Response resolve and Unmarshal methods.These should be `fs` methods.

Reply:
I changed this based on the code review from @mrahatm

Reply:
I see. Added a comment in that discussion - I'll defer to his final opinion.

Reply:
He can be wrong too. If someone suggests to change something, they have to explain why they think that it will be better with the suggested changes. Then if you agree with the explanation, you need to change your code. Otherwise you need to continue a discussion to find the right way.

Reply:
I did agree with @mrahatm since all we did is indeed initializing the connection, not the file system

Reply:
+1, no one dictates how something should be done. Suggestions should be backed up by reasoning and everyone involved needs to reach a consensus based on the provided reasons. If something is ambiguous (i.e. it comes down to a style choice, or all options seem equally valid), the author of the code has final say.